sources/mdb_iamproxy.c

```diff
@@ -0,0 +1,229 @@
+
+/*
+ * Odyssey.
+ *
+ * Scalable PostgreSQL connection pooler.
+ */
+
+#include <odyssey.h>
+#include <stdint.h>
+#include <malloc.h>
+#include <sys/poll.h>
+#include <sys/un.h>
+#include <sys/socket.h>
+
+/*CONNECTION CALLBACK TYPES*/
+#define MDB_IAMPROXY_CONN_ERROR -1
+#define MDB_IAMPROXY_CONN_TIMEOUT -1
+#define MDB_IAMPROXY_CONN_ACCEPTED 0
+#define MDB_IAMPROXY_CONN_DENIED -1
+
+#define MDB_IAMPROXY_RES_ERROR -1
+#define MDB_IAMPROXY_RES_OK 0
+
+/*AUTHENTICATION TIMEOUT LIMIT*/
+#define MDB_IAMPROXY_BYTE_SIZE 8
+#define MDB_IAMPROXY_DEFAULT_HEADER_SIZE 8
+#define MDB_IAMPROXY_DEFAULT_CNT_CONNECTIONS 1
+
+#define MDB_IAMPROXY_DEFAULT_CONNECTION_TIMEOUT 1000
+#define MDB_IAMPROXY_DEFAULT_RECEIVING_TIMEOUT 1000
+
+/*PAM SOCKET FILE*/
+#define MDB_IAMPROXY_DEFAULT_SOCKET_FILE \
+	"/var/run/iam-auth-proxy/iam-auth-proxy.sock" // PAM SOCKET FILE place
+
+int mdb_iamproxy_recv_from_socket(int socket_fd, char *msg_body)
+{
+	/*GET COMMON MSG INFO AND ALLOCATE RESOURCES*/
+	uint8_t buffer[8];
+	uint64_t body_size = 0;
+	uint64_t received = 0;
+
+	/*RECEIVE HEADER*/
+	while (received < MDB_IAMPROXY_DEFAULT_HEADER_SIZE) {
+		int rt = recv(socket_fd, buffer + received,
+			      MDB_IAMPROXY_DEFAULT_HEADER_SIZE - received, 0);
+		if (rt < 0) {
+			return MDB_IAMPROXY_CONN_ERROR;
+		}
+		received += rt;
+		;
+	}
+	for (int i = 0; i < MDB_IAMPROXY_DEFAULT_HEADER_SIZE; ++i) {
+		body_size |=
+			(((uint64_t)buffer[i]) << (i * MDB_IAMPROXY_BYTE_SIZE));
+	}
+
+	/*RECEIVE BODY*/
+	received = 0;
+	while (received < body_size) {
+		int rt = recv(socket_fd, msg_body + received,
+			      body_size - received, 0);
+		if (rt < 0) {
+			return MDB_IAMPROXY_CONN_ERROR;
+		}
+		received += rt;
+	}
+
+	return MDB_IAMPROXY_CONN_ACCEPTED;
+}
+
+int mdb_iamproxy_send_to_socket(int socket_fd, const char *send_msg)
+{
+	/*GET COMMON MSG INFO AND ALLOCATE BUFFER*/
+	int32_t send_result = MDB_IAMPROXY_RES_OK;
+	uint64_t body_size =
+		strlen(send_msg) +
+		1; // stores size of message (add one byte for 'c\0')
+	uint64_t current_body_size = body_size;
+	uint64_t msg_size = sizeof(body_size) + body_size;
+	uint64_t sent = 0; // stores byte-size of sended info
+	char *msg = (char *)calloc(
+		msg_size, sizeof(*msg)); // allocate memory for msg buffer
```

Comment:
sizeof(char) == 1 by C standard

Reply:
Yes, but what's incorrect here? i think that it's common to allocate dynamic memory in that way 

type *mem = calloc(n, sizeof(*mem)); its easily to read, and remove nubmers/random (for common code reader) constants from code

But if necessary, I can fix it